pkg/serviceaccount/openidmetadata.go

```diff
@@ -0,0 +1,283 @@
+	}, nil
+}
+
+// openIDMetadata provides a minimal subset of OIDC provider metadata:
+// https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
+type openIDMetadata struct {
+	Issuer string `json:"issuer"` // REQUIRED in OIDC; meaningful to relying parties.
+	// TODO(mtaufen): Since our goal is compatibility for relying parties that
+	// need to validate ID tokens, but do not need to initiate login flows,
+	// and since we aren't sure what to put in authorization_endpoint yet,
+	// we will omit this field until someone files a bug.
+	// AuthzEndpoint string   `json:"authorization_endpoint"`                // REQUIRED in OIDC; but useless to relying parties.
+	JWKSURI       string   `json:"jwks_uri"`                              // REQUIRED in OIDC; meaningful to relying parties.
+	ResponseTypes []string `json:"response_types_supported"`              // REQUIRED in OIDC
+	SubjectTypes  []string `json:"subject_types_supported"`               // REQUIRED in OIDC
+	SigningAlgs   []string `json:"id_token_signing_alg_values_supported"` // REQUIRED in OIDC
+}
+
+// openIDConfigJSON returns the JSON OIDC Discovery Doc for the service
+// account issuer.
+func openIDConfigJSON(iss, jwksURI string, keys []interface{}) ([]byte, error) {
+	keyset, errs := publicJWKSFromKeys(keys)
+	if errs != nil {
+		return nil, errs
+	}
+
+	metadata := openIDMetadata{
+		Issuer:        iss,
+		JWKSURI:       jwksURI,
+		ResponseTypes: []string{"id_token"}, // Kubernetes only produces ID tokens
+		SubjectTypes:  []string{"public"},   // https://openid.net/specs/openid-connect-core-1_0.html#SubjectIDTypes
+		SigningAlgs:   getAlgs(keyset),      // REQUIRED by OIDC
+	}
+
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal service account issuer metadata: %v", err)
+	}
+
+	return metadataJSON, nil
+}
+
+// openIDKeysetJSON returns the JSON Web Key Set for the service account
+// issuer's keys.
+func openIDKeysetJSON(keys []interface{}) ([]byte, error) {
+	keyset, errs := publicJWKSFromKeys(keys)
+	if errs != nil {
+		return nil, errs
+	}
+
+	keysetJSON, err := json.Marshal(keyset)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal service account issuer JWKS: %v", err)
+	}
+
+	return keysetJSON, nil
+}
+
+func getAlgs(keys *jose.JSONWebKeySet) []string {
+	algs := sets.NewString()
+	for _, k := range keys.Keys {
+		algs.Insert(k.Algorithm)
+	}
+	// Note: List returns a sorted slice.
+	return algs.List()
+}
+
+// publicJWKSFromKeys constructs a JSONWebKeySet from a list of keys. The key
+// set will only contain the public keys associated with the input keys.
+func publicJWKSFromKeys(in []interface{}) (*jose.JSONWebKeySet, errors.Aggregate) {
+	// Decode keys into a JWKS.
+	var keys jose.JSONWebKeySet
+	var errs []error
+	for i, key := range in {
+		var pubkey *jose.JSONWebKey
+		var err error
+
+		switch k := key.(type) {
+		case interface {
+			Public() crypto.PublicKey
+		}:
```

Comment:
this would be more readable as a local interface above this method

```go
type publicKeyGetter interface {
  Public() crypto.PublicKey
}
```

Reply:
Done.